bin/requirements.rb

```diff
@@ -0,0 +1,187 @@
+  end
+
+  def parse
+    filenames.each do |fn|
+      # the list of requirements-files can have items commented out - ignore those
+
+      mod = module_name_from_filename(fn)
+      IO.read(fn).split("\n").each do |line|
+        lib, ver = parse_line(line)
+        next unless lib
+
+        # ===
+        # === business rules
+        # ===
+
+        # skip git libraries. git>= line from vsphere gave us problems
+        next if lib.start_with?("git")
+
+        # Do not version system provided libraries: requests, requests[security], pyspnego[kerberos]
+        ver = "" if lib.match?(/^requests($|\[)/) || lib.start_with?("pyspnego")
+
+        final[lib] ||= {}
+        (final[lib][ver] ||= []) << mod
+      end
+    end
+
+    self
+  end
+
+  def output
+    result = final.flat_map do |lib, vers|
+      # consolidate multiple versioning rules
+      if vers.size > 1
+        max_key, *all_keys = vers.keys
+        all_keys.each do |alt|
+          higher, lower, conflict = version_compare(alt, max_key)
+          # There is a conflict when we have something like >=2.0 and ==1.0
+          #
+          # We are just displaying all comparisons/winners to verify comparisons (skipping when a blank is provided)
+          $stderr.puts "#{lib}: #{higher} > #{lower} #{"CONFLICT" if conflict}" if lower != ""
+          vers[higher].concat(vers.delete(lower))
+          max_key = higher
+        end
+      end
+
+      vers.map do |(ver, modules)|
+        # if we pass in a previous requirements.txt, lets not mention it
+        modules.delete("legacy") if modules.size > 1
+        "#{lib}#{ver} # #{modules.join(", ")}"
+      end
+    end.sort.join("\n")
+
+    puts result
+  end
+
+  private
+
+  def module_name_from_filename(fn)
+    if non_modules.include?(fn)
+      "legacy"
+    else
+      fn.gsub(%r{.*ansible_collections/}, "")
+        .gsub(%r{/requirements.*}, "")
+    end
+  end
+
+  def parse_line(line)
+    line.downcase!
+    # TODO: do we want to keep legacy comments? Only useful for our requirements.txt file
+    line.gsub!(/#.*/, "")
+    line.strip!
+    return if line.empty?
+
+    # some libraries list "python" instead of "python_version"
+    line.gsub!(/^python /, "python_version")
+    # change version 5+ to >=5
+    line.gsub!(/\([0-9.]*\)[+]/, '>=\1')
+    line.gsub!("= ", "=")
+    # ignore line where a module requires a minumum python version (assumption here)
+    return if line.match?(/^python_version/)
+    # ignore package requirements for older version of pythons (assumption here)
+    return if line.match?(/python_version ?[=<]/)
+
+    # split on first space (or =)
+    # version can have multiple spaces
+    lib, ver = line.match(/([^ >=]*) ?(.*)/).captures
+    # have already normalized for lowercase
+    # normalize library name with dash. all are the same
+    lib.gsub!(/[-_.]+/, "-")
+    ver ||= ""
+
+    [lib, ver]
+  end
+
+  # return highest, lowest for version comparison
+  def version_compare(a, b)
```

Comment:
Can we use Gem::Version object here to do the heavy lifting or are the requirements not quite the same?  e.g. something like:

```suggestion
  def version_compare(a, b)
    [Gem::Version.new(a), Gem::Version.new(b)].sort.reverse
  end
```
```

Reply:
or perhaps Gem::Requirement...

```ruby
a = Gem::Requirement.new(">=4.5.0")
a.requirements
# => [[">=", Gem::Version.new("4.5.0")]]

a = Gem::Requirement.new([">=4.5.0", ">=1.2.3"])
a.requirements
# => [[">=", Gem::Version.new("4.5.0")], [">=", Gem::Version.new("1.2.3")]]
```

Reply:
Python and ruby seem very close here - https://pip.pypa.io/en/stable/reference/requirements-file-format/

I do like the idea of requirements instead of doing this ourselves.
But in the end of the day, there are only 2 instances where it is not a simple "merge no dependency with a dependency"

And I'm not sure that we want to invent a full requirements.txt parser in ruby to handle the `,` and `;` and all the wonderful `python_version` stuff.